cpp/src/parquet/encryption/encryption_internal.cc

```diff
@@ -438,56 +462,102 @@ std::shared_ptr<AesDecryptor> AesDecryptor::Make(
   return decryptor;
 }
 
-int AesDecryptor::CiphertextSizeDelta() { return impl_->ciphertext_size_delta(); }
-
-int AesDecryptor::AesDecryptorImpl::GcmDecrypt(const uint8_t* ciphertext,
-                                               int ciphertext_len, const uint8_t* key,
-                                               int key_len, const uint8_t* aad,
-                                               int aad_len, uint8_t* plaintext) {
-  int len;
-  int plaintext_len;
+int AesDecryptor::PlaintextLength(int ciphertext_len) const {
+  return impl_->PlaintextLength(ciphertext_len);
+}
 
-  uint8_t tag[kGcmTagLength];
-  memset(tag, 0, kGcmTagLength);
-  uint8_t nonce[kNonceLength];
-  memset(nonce, 0, kNonceLength);
+int AesDecryptor::CiphertextLength(int plaintext_len) const {
+  return impl_->CiphertextLength(plaintext_len);
+}
 
+int AesDecryptor::AesDecryptorImpl::GetCiphertextLength(
+    ::arrow::util::span<const uint8_t> ciphertext) const {
   if (length_buffer_length_ > 0) {
+    if (ciphertext.size() < static_cast<size_t>(length_buffer_length_)) {
```

Comment:
It took me a while to understand this line. Perhaps it is better to explicitly use `kBufferSizeLength` here as line 484 to 486 have assumed this length is 4 bytes.

Reply:
Yes good point, that also confused me a bit. I've changed that and added a comment to hopefully make this more readable